lib/sqlitex/query.ex

```diff
@@ -1,87 +1,48 @@
-            column_names: [],
-            column_types: [],
-            database: nil,
-            into: [],
-            raw_data: [],
-            sql: nil,
-            statement: nil
+  alias Sqlitex.Statement
 
-  def query(db, sql, opts \\ []) do
-    pipe_matching {:ok, _},
-      {:ok, %Sqlitex.Query{bindings: bindings_from_opts(opts), into: into_from_opts(opts), database: db, sql: sql}}
-        |> prepare
-        |> bind_values
-        |> column_types
-        |> column_names
-        |> fetch_data
-        |> into_rows
-  end
-
-  defp bind_values({:ok, %Sqlitex.Query{bindings: bindings, statement: statement}=query}) do
-    case :esqlite3.bind(statement, bindings) do
-      {:error,_}=error -> error
-      :ok -> {:ok, query}
-    end
-  end
+  @doc """
+  Runs a query and returns the results.
 
-  defp bindings_from_opts(opts), do: opts |> Keyword.get(:bind, []) |> translate_bindings
+  ## Parameters
 
-  defp column_names({:ok, %Sqlitex.Query{statement: statement}=query}) do
-    case :esqlite3.column_names(statement) do
-      {:error, _}=other -> other
-      names -> {:ok, %Sqlitex.Query{query | column_names: names}}
-    end
-  end
+  * `db` - A sqlite database.
+  * `sql` - The query to run as a string.
+  * `opts` - Options to pass into the query.  See below for details.
 
-  defp column_types({:ok, %Sqlitex.Query{statement: statement}=query}) do
-    case :esqlite3.column_types(statement) do
-      {:error, _}=other -> other
-      types -> {:ok, %Sqlitex.Query{query | column_types: types}}
-    end
-  end
+  ## Options
 
-  defp datetime_to_string({{yr, mo, da}, {hr, mi, se, usecs}}) do
-    [zero_pad(yr, 4), "-", zero_pad(mo, 2), "-", zero_pad(da, 2), " ", zero_pad(hr, 2), ":", zero_pad(mi, 2), ":", zero_pad(se, 2), ".", zero_pad(usecs, 6)]
-    |> Enum.join
-  end
+  * `bind` - If your query has parameters in it, you should provide the options
+    to bind as a list.
+  * `into` - The collection to put results into.  This defaults to a list.
 
-  defp fetch_data({:ok, %Sqlitex.Query{statement: statement}=query}) do
-    case :esqlite3.fetchall(statement) do
-      {:error,_}=other -> other
-      raw_data -> {:ok, %Sqlitex.Query{query | raw_data: raw_data}}
-    end
+  ## Returns
+  * {:ok, results} on success
+  * {:error, _} on failure.
+  """
+  def query(db, sql, opts \\ []) do
+    pipe_with &pipe_ok/2,
+      Statement.prepare(db, sql)
+      |> Statement.bind_values(Dict.get(opts, :bind, []))
+      |> Statement.fetch_all(Dict.get(opts, :into, []))
   end
 
-  defp into_from_opts(opts), do: Keyword.get(opts, :into, [])
+  @doc """
+  Same as `query/3` but raises an error on error.
 
-  defp prepare({:ok, %Sqlitex.Query{sql: sql, database: database}=query}) do
-    case :esqlite3.prepare(sql, database) do
-      {:ok, statement} -> {:ok, %Sqlitex.Query{query | statement: statement}}
-      other -> other
-    end
+  Returns the results otherwise.
+  """
+  def query!(db, sql, opts \\ []) do
```

Comment:
Just realised that is not the best example, because `Sqlitex.query` doesn't return a struct of `{:ok, results}`, it just returns `results`.  I suppose there's less need for it then, though I still think it could be useful for piping - if you used the normal query in a pipe then you could end up passing an error tuple into a function that's not expecting one.  Would either cause an odd looking non-match, or maybe even some unexpected results depending on how the function was coded...

I assume you'd get some typespec related warnings (assuming Sqlitex had typespecs) if you were piping `Sqlitex.query` into a function that wasn't expecting to deal with `{:error, x}` tuples too.